src/metabase/api/dashboard.clj

```diff
@@ -580,37 +646,65 @@
+  [:map
+   ;; id can be negative, it indicates a new card and BE should create them
+   [:id   ms/Int]
+   [:name ms/NonBlankString]])
+
+(defn- update-dashboard-with-tabs!
+  [dashboard new-tabs new-cards]
+  (let [{:keys [temp->real-tab-ids
+                deleted-tab-ids]}  (do-update-tabs! dashboard (:ordered_tabs dashboard) new-tabs)
+        current-cards              (if (seq deleted-tab-ids)
+                                       (remove (fn [card]
+                                                 (contains? deleted-tab-ids (:dashboard_tab_id card)))
+                                               (:ordered_cards dashboard))
+                                       (:ordered_cards dashboard))
+        new-cards                  (cond->> new-cards
+                                     ;; fixup the temporary tab ids with the real ones
+                                     (seq temp->real-tab-ids)
+                                     (map (fn [card]
+                                            (if-let [real-tab-id (get temp->real-tab-ids (:dashboard_tab_id card))]
+                                              (assoc card :dashboard_tab_id real-tab-id)
+                                              card)))
+                                     ;; if there are deleted tabs, we need to remove the dashcards from those tabs
+                                     (seq deleted-tab-ids)
+                                     (remove #(contains? deleted-tab-ids (:dashboard_tab_id %))))]
+    (do-update-dashcards! dashboard current-cards new-cards)))
+
 (api/defendpoint PUT "/:id/cards"
-  "Update `Cards` on a Dashboard. Request body should have the form:
-
-    {:cards [{:id                 ... ; DashboardCard ID
-              :size_x             ...
-              :size_y             ...
-              :row                ...
-              :col                ...
-              :parameter_mappings ...
-              :series             [{:id 123
-                                    ...}]}
-             ...]}"
-  [id :as {{:keys [cards]} :body}]
+  "Update `Cards` and `Tabs` on a Dashboard. Request body should have the form:
+
+    {:cards        [{:id                 ... ; DashboardCard ID
+                     :size_x             ...
+                     :size_y             ...
+                     :row                ...
+                     :col                ...
+                     :parameter_mappings ...
+                     :series             [{:id 123
+                                           ...}]}
+                     ...]
+     :ordered_tabs [{:id       ... ; DashboardTab ID
+                     :name     ...}]}"
+  [id :as {{:keys [cards ordered_tabs]} :body}]
   {id    ms/PositiveInt
-   cards [:sequential UpdatedDashboardCard]}
+   cards [:maybe (ms/maps-with-unique-key [:sequential UpdatedDashboardCard] :id)]
+   ordered_tabs  [:maybe (ms/maps-with-unique-key [:sequential UpdatedDashboardTab] :id)]}
```

Comment:
we allow a dashboard to have no tabs. this partly responsible for dashboard created pre-47 and also there are path way to create them in 47.

Reply:
If a dashboard has tabs persisted, then the FE sends a request with `ordered_tabs=nil`, we'd still need to `do-update-dashcards!` to take care of the deleted tabs. Is there any difference in meaning between `ordered_tabs=nil` and `ordered_tabs=[]`? If not, then they should be treated the same here. Then you wouldn't need the branching in `(if-not (nil? ordered-tabs) ...) or `(when-not (nil? ordered_tabs) ...)`.

Reply:
I originally kept that behavior of `ordered_tabs` is nullable just because I don't want to update all the current tests to have that key, but I guess we've already done a refactor so it's worth keeping everything up to date.

I've updated to make `ordered_tabs` is not nullable but can be empty : 7d6e9b1fa0